wgsl/index.bs

```diff
@@ -5188,6 +5188,77 @@ A phony-assignment is useful for:
   </xmp>
 </div>
 
+### Complex assignment ### {#complex-assignment}
+
+A <dfn noexport>complex assignment</dfn> is a form of syntactic sugar that combines an operator and an assignment.
+
+<div class='syntax' noexport='true'>
+  <dfn for=syntax>complex_assignment_statement</dfn> :
+
+    | [=syntax/unary_expression=] [=syntax/plus_equal=] [=syntax/short_circuit_or_expression=]
+
+    | [=syntax/unary_expression=] [=syntax/minus_equal=] [=syntax/short_circuit_or_expression=]
+
+    | [=syntax/unary_expression=] [=syntax/times_equal=] [=syntax/short_circuit_or_expression=]
+
+    | [=syntax/unary_expression=] [=syntax/division_equal=] [=syntax/short_circuit_or_expression=]
+
+    | [=syntax/unary_expression=] [=syntax/modulo_equal=] [=syntax/short_circuit_or_expression=]
+
+    | [=syntax/unary_expression=] [=syntax/and_equal=] [=syntax/short_circuit_or_expression=]
+
+    | [=syntax/unary_expression=] [=syntax/or_equal=] [=syntax/short_circuit_or_expression=]
+
+    | [=syntax/unary_expression=] [=syntax/xor_equal=] [=syntax/short_circuit_or_expression=]
+
+    | [=syntax/unary_expression=] [=syntax/shift_right_equal=] [=syntax/short_circuit_or_expression=]
+
+    | [=syntax/unary_expression=] [=syntax/shift_left_equal=] [=syntax/short_circuit_or_expression=]
+</div>
+
+It is fully defined by the equivalent code that it expands into:
+
+<table class='data'>
+  <thead>
+    <tr><th>Statement<th>Expansion
+  </thead>
+  <tr>
+    <td class="nowrap">|e1| += |e2|
+    <td>|e1| = |e1| + |e2|
+<tr>
+    <td class="nowrap">|e1| -= |e2|
+    <td>|e1| = |e1| - |e2|
+<tr>
+    <td class="nowrap">|e1| *= |e2|
+    <td>|e1| = |e1| * |e2|
+<tr>
+    <td class="nowrap">|e1| /= |e2|
+    <td>|e1| = |e1| / |e2|
+<tr>
+    <td class="nowrap">|e1| %= |e2|
+    <td>|e1| = |e1| % |e2|
+<tr>
+    <td class="nowrap">|e1| &= |e2|
+    <td>|e1| = |e1| & |e2|
+<tr>
+    <td class="nowrap">|e1| |= |e2|
+    <td>|e1| = |e1| | |e2|
+<tr>
+    <td class="nowrap">|e1| ^= |e2|
+    <td>|e1| = |e1| ^ |e2|
+<tr>
+    <td class="nowrap">|e1| >>= |e2|
+    <td>|e1| = |e1| >> |e2|
+<tr>
+    <td class="nowrap">|e1| <<= |e2|
+    <td>|e1| = |e1| << |e2|
+</table>
+
+Note: in all of the cases above, |e1| is guaranteed by a combination of the syntax and typing rules to be composed of some optional sequence of `*` and `&` followed by an identifier.
```

Comment:
I dislike that the constraints on the LHS are unlike for regular assignment.

Instead, I would rather that the LHS is a reference that could be on a regular assignment, and whose store type allows the operation.

And then the translation for

>  LHS _op_= RHS

would be

>   { let p = &LHS; 
>   *p = *p _op_ RHS }

Then there is no concern for issuing side effects multiple times.

Reply:
Hm..  Except for the component-of-a-vector case.


          var v = vec4<f32>();
          v.y += 1.0;

We can't take the address of a component of a vector.  :-(

Reply:
The only difference in syntactic constraints with regular assignment is that `_` is not allowed on the LHS side.
The meaningful constraints are in terms of type, and they flow naturally from the fact that the expression on the LHS must also be a valid left-operand to the operator.

My hope was that by making complex assignments mere syntactic sugar I could both minimize implementers' burden, and make it conceptually easy for programmers.

Reply:
> The only difference in syntactic constraints with regular assignment is that _ is not allowed on the LHS side.

Maybe I was confused.

This PR has the LHS as unary_expression.  That might work, but I'd see it more clearly as similar to regular assignment if it uses lhs_expression.  I kind of got lost tracing back through unary_expression to see what it expands to.